### PR TITLE
Split textDocument/formatting TextEdit with diff

### PR DESCRIPTION
**Pull request by Jesse-Bakker**
#7044

### CODE REVIEW

File: crates/rust-analyzer/Cargo.toml [hunk 1]
```diff
@@ -17,6 +17,7 @@ path = "src/bin/main.rs"
 [dependencies]
 anyhow = "1.0.26"
 crossbeam-channel = "0.5.0"
+dissimilar = "1.0.2"
```
Comment (matklad) [inline crates/rust-analyzer/Cargo.toml:20]: If you could send a patch for https://github.com/rust-analyzer/expect-test to use dissimilar as well, that'd be sweet :)

File: crates/rust-analyzer/src/diff.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+            }
+        }
+    }
+    if let Some(delete_start) = pending_delete_start {
+        builder.delete(TextRange::new(delete_start, pos));
+    }
+    builder.finish()
+}
+
+pub(crate) fn diff(left: &str, right: &str) -> TextEdit {
```
Comment (matklad) [inline crates/rust-analyzer/src/diff.rs:44]: Let's move this fn to top, as per https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#order-of-items

File: crates/rust-analyzer/src/diff.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+                        pos += TextSize::of(text);
+                    }
+                    Chunk::Insert(text) => {
+                        builder.insert(pos, text.into());
+                        pos += TextSize::of(text);
+                    }
+                }
+            }
+        }
+    }
```
Comment (matklad) [inline crates/rust-analyzer/src/diff.rs:37]: Would using  `chunks.peekable` help here? Seems like we could detect delet-insert pairs that way.

**Review by matklad — COMMENTED** [review]
:heart_eyes:

File: crates/rust-analyzer/src/diff.rs (hunk 3 above)
Comment (Jesse-Bakker) [reply crates/rust-analyzer/src/diff.rs:37]: I see two ways how that could work. One is to move ownership of the iterator out of the loop and skip the the insert iteration if it is peeked, but that gets pretty gnarly. I tried that, and the general logic gets a bit simpler, but the control flow gets difficult to follow. The other way would be to set a variable that an iteration should be skipped, which would look very similar to the current implementation. 

A solution that's maybe clearer would be to introduce an enum similar to Chunk, but with Replace, then Iterator::map chunks to that and Itertools::coalesce the result and map that to Indel

Comment (matklad) [reply crates/rust-analyzer/src/diff.rs:37]: Hm, 

```rust
fn textedit_from_chunks(chunks: Vec<dissimilar::Chunk>) -> TextEdit {
    let mut builder = TextEdit::builder();
    let mut pos = TextSize::default();

    let mut chunks = chunks.into_iter().peekable();
    while let Some(ch) = chunks.next() {
        if let (Chunk::Delete(del), Some(&Chunk::Insert(ins))) = (ch, chunks.peek()) {
            chunks.next().unwrap();
            let del_len = TextSize::of(del);
            builder.replace(TextRange::at(pos, del_len), ins.into());
            pos += del_len;
            continue;
        }

        match ch {
            Chunk::Equal(text) => {
                pos += TextSize::of(text);
            }
            Chunk::Delete(del) => {
                let del_len = TextSize::of(del);
                builder.delete(TextRange::at(pos, del_len));
                pos += del_len;
            }
            Chunk::Insert(text) => {
                builder.insert(pos, text.into());
            }
        }
    }

    builder.finish()
}
```

seems a bit simpler to me....

File: crates/rust-analyzer/src/diff.rs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+                    Chunk::Equal(text) => {
+                        pos += TextSize::of(text);
+                    }
+                    Chunk::Delete(text) => {
+                        pending_delete_start = Some(pos);
+                        pos += TextSize::of(text);
+                    }
+                    Chunk::Insert(text) => {
+                        builder.insert(pos, text.into());
+                        pos += TextSize::of(text);
```
Comment (matklad) [inline crates/rust-analyzer/src/diff.rs:32]:
```suggestion
```

Looking at this again, this feels like a bug?

File: crates/rust-analyzer/Cargo.toml (hunk 1 above)
Comment (Jesse-Bakker) [reply crates/rust-analyzer/Cargo.toml:20]: Will do

File: crates/rust-analyzer/src/diff.rs (hunk 3 above)
Comment (Jesse-Bakker) [reply crates/rust-analyzer/src/diff.rs:37]: Yes, you are right, that is simpler. I probably overcomplicated it a bit when I tried it that way :)

File: crates/rust-analyzer/src/diff.rs (hunk 4 above)
Comment (Jesse-Bakker) [reply crates/rust-analyzer/src/diff.rs:32]: Yep

**Pull request merged**